common/src/main/java/com/wynntils/services/hint/HintService.java

```diff
@@ -0,0 +1,147 @@
+import java.io.Reader;
+import java.lang.reflect.Type;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Random;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import net.minecraft.ChatFormatting;
+import net.minecraft.client.KeyMapping;
+import net.minecraft.client.resources.language.I18n;
+import net.minecraft.network.chat.ClickEvent;
+import net.minecraft.network.chat.Component;
+import net.minecraft.network.chat.HoverEvent;
+import net.minecraft.network.chat.MutableComponent;
+
+public class HintService extends Service {
+    private static final String DEFAULT_LANGUAGE = "en_us";
+    private static final String UNBOUND_KEY = "key.keyboard.unknown";
+    private static final Random RANDOM = new Random();
+
+    private static final Pattern VARIABLE_PATTERN = Pattern.compile("\\{(?<action>[^:}]+):(?<value>[^}]+)\\}");
+
+    private List<Map<String, String>> hints = new ArrayList<>();
+
+    public HintService() {
+        super(List.of());
+    }
+
+    @Override
+    public void registerDownloads(DownloadRegistry registry) {
+        registry.registerDownload(UrlId.DATA_STATIC_HINTS).handleReader(this::handleHintMessages);
+    }
+
+    private void handleHintMessages(Reader reader) {
+        Type type = new TypeToken<List<Map<String, String>>>() {}.getType();
+        hints = WynntilsMod.GSON.fromJson(reader, type);
+    }
+
+    public void sendHint() {
+        String languageCode = McUtils.mc().getLanguageManager().getSelected();
+
+        if (hints.isEmpty()) return;
+
+        int hintNum = RANDOM.nextInt(hints.size());
+        Map<String, String> currentHint = hints.get(hintNum);
+        String hint;
+
+        if (currentHint.containsKey(languageCode)) {
+            hint = currentHint.get(languageCode);
+        } else {
+            hint = currentHint.getOrDefault(DEFAULT_LANGUAGE, "");
+        }
+
+        if (hint.isEmpty()) {
+            WynntilsMod.warn("Hint " + hintNum + " has no " + DEFAULT_LANGUAGE + " translation.");
+            return;
+        }
+
+        MutableComponent component = formatHint(hint);
+
+        if (component.equals(Component.empty())) return;
+
+        McUtils.sendMessageToClientWithPillHeader(component);
+    }
+
+    private MutableComponent formatHint(String hint) {
+        MutableComponent component = Component.empty();
+
+        Matcher matcher = VARIABLE_PATTERN.matcher(hint);
+        List<String> parts = new ArrayList<>();
+        int lastEnd = 0;
+
+        while (matcher.find()) {
+            if (matcher.start() > lastEnd) {
+                parts.add(hint.substring(lastEnd, matcher.start()));
+            }
+
+            parts.add(matcher.group());
+            lastEnd = matcher.end();
+        }
+
+        if (lastEnd < hint.length()) {
+            parts.add(hint.substring(lastEnd));
+        }
+
+        for (String part : parts) {
+            Matcher variableMatcher = VARIABLE_PATTERN.matcher(part);
```

Comment:
Just for code clarity, i'd store matches in a Pair<ActionType, String>, then just iterate and not rematch.

Reply:
Also, wouldn't it be better to use StyledText in the hint static storage / here? That would allow us a bit more customization. You can keep the default formatting as part of the code here, eg. as a formatted parent type. Then, if there is any formatting appended to the hint, it'll be applied (and it'll inherit everything else).